### PR TITLE
Reworked camera

### PR DESCRIPTION
**Pull request by LeoWikstrom**
Player turns in attacking direction

### CODE REVIEW

File: StortSpelProjekt/Engine/src/Renderer/Camera/BaseCamera.cpp [hunk 1]
```diff
@@ -7,7 +7,7 @@ BaseCamera::BaseCamera(DirectX::XMVECTOR position, DirectX::XMVECTOR direction)
 	m_EyeVector = position;
 	m_DirectionVector = direction;
 	m_UpVector = DirectX::XMVectorSet(0.0f, 1.0f, 0.0f, 0.0f);
-	m_RightVector = DirectX::XMVector3Cross(m_DirectionVector, m_UpVector);
+	m_RightVector = DirectX::XMVector3Cross(m_UpVector, m_DirectionVector);
```
Comment (fjz345) [inline StortSpelProjekt/Engine/src/Renderer/Camera/BaseCamera.cpp:11]: You changed rightvector from rightvector to be left vector.

File: StortSpelProjekt/Engine/src/Renderer/Camera/BaseCamera.cpp [hunk 2]
```diff
@@ -36,7 +36,7 @@ void BaseCamera::SetPosition(float x, float y, float z)
 void BaseCamera::SetDirection(float x, float y, float z)
 {
 	m_DirectionVector = DirectX::XMVectorSet(x, y, z, 0.0f);
-	m_RightVector = DirectX::XMVector3Cross(m_DirectionVector, m_UpVector);
+	m_RightVector = DirectX::XMVector3Cross(m_UpVector, m_DirectionVector);
 }
```
Comment (fjz345) [inline StortSpelProjekt/Engine/src/Renderer/Camera/BaseCamera.cpp:40]: Same as above.

File: StortSpelProjekt/Engine/src/Renderer/Camera/PerspectiveCamera.cpp [hunk 3]
```diff
@@ -85,7 +85,7 @@ void PerspectiveCamera::SetMovement(float x, float y, float z)
 
 void PerspectiveCamera::updateCameraMovement(double dt)
 {
-	m_RightVector = DirectX::XMVector3Cross(m_DirectionVector, m_UpVector);
+	m_RightVector = DirectX::XMVector3Cross(m_UpVector, m_DirectionVector);
```
Comment (fjz345) [inline StortSpelProjekt/Engine/src/Renderer/Camera/PerspectiveCamera.cpp:89]: Same as above.

File: StortSpelProjekt/Game/src/Gamefiles/Components/MeleeComponent.cpp [hunk 4]
```diff
@@ -112,6 +113,14 @@ void component::MeleeComponent::Attack()
...
 		m_Cooldown = true;
 		m_TimeSinceLastAttackCheck = 0;
+
+		float3 forward = m_pParent->GetComponent<component::CameraComponent>()->GetCamera()->GetDirectionFloat3();
+
+		double angle = std::atan2(forward.x, forward.z);
+		int angleDegrees = EngineMath::convertToWholeDegrees(angle);
+		angleDegrees = (angleDegrees + 360) % 360;
+		m_pParent->GetComponent<component::PlayerInputComponent>()->SetAngleToTurnTo(angleDegrees);
+		m_pParent->GetComponent<component::PlayerInputComponent>()->SetAttacking();
```
Comment (fjz345) [inline StortSpelProjekt/Game/src/Gamefiles/Components/MeleeComponent.cpp:124]: Feels like this code does not belong in melee component.

File: StortSpelProjekt/Game/src/Gamefiles/Components/PlayerInputComponent.cpp [hunk 5]
```diff
@@ -15,24 +15,22 @@ component::PlayerInputComponent::PlayerInputComponent(Entity* parent, unsigned i
...
-	m_CameraDistance = sqrt(m_Zoom * 4 * m_Zoom * 4 + m_Zoom * m_Zoom);
+	m_MovementSpeed = 10.0f;
 
 	m_pCamera = nullptr;
 	m_pTransform = nullptr;
 	m_pCC = nullptr;
 
 	m_Dashing = false;
 	m_DashReady = true;
+	m_Attacking = false;
```
Comment (fjz345) [inline StortSpelProjekt/Game/src/Gamefiles/Components/PlayerInputComponent.cpp:33]: I would put this variable in meleeComponent and if you need to check it, ask meleeComponent if it's attacking.

File: StortSpelProjekt/Game/src/Gamefiles/Components/PlayerInputComponent.cpp [hunk 6]
```diff
@@ -100,17 +102,36 @@ void component::PlayerInputComponent::RenderUpdate(double dt)
...
+		cameraPos.y = min(max(cameraPos.y + m_RotateY * 100.0f * static_cast<float>(dt), -80.0f), 80.0f);
+		m_pCamera->SetPosition(cameraPos.x, cameraPos.y, cameraPos.z);
+
+		m_RotateX *= 100.0f * dt;
+
+		cameraDir = {
+			cameraDir.x * cos(m_RotateX) + cameraDir.z * sin(m_RotateX),
+			playerPosition.y - cameraPos.y + (static_cast<float>(m_pParent->GetComponent<component::ModelComponent>()->GetModelDim().y) * m_pTransform->GetScale().y * 0.5f) + 1.0f,
+			-cameraDir.x * sin(m_RotateX) + cameraDir.z * cos(m_RotateX)
+		};
```
Comment (fjz345) [inline StortSpelProjekt/Game/src/Gamefiles/Components/PlayerInputComponent.cpp:120]: Comment more overall, this section I would like a comment on this big chunk of code or put it in a function with a descriptive name.

File: StortSpelProjekt/Game/src/Gamefiles/Components/PlayerInputComponent.cpp [hunk 7]
```diff
@@ -119,6 +140,21 @@ void component::PlayerInputComponent::RenderUpdate(double dt)
 
 	m_DashTimer += dt;
 	m_DashReady = m_DashTimer > 1.5;
+	
+	/* Update model rotation */
+	double3 rot = m_pCC->GetRotationEuler();
+	int angle = EngineMath::convertToWholeDegrees(rot.y);
+	int mod = EngineMath::convertToWholeDegrees(rot.x) / 180;
+
+	angle = (mod * 360 + (mod * -2 + 1) * ((angle + 360) % 360) + mod * 180) % 360;
```
Comment (fjz345) [inline StortSpelProjekt/Game/src/Gamefiles/Components/PlayerInputComponent.cpp]: I know it's not in your task, but this renderUpdate() function is getting quite long. Might want to add some stuff in private functions to make it easier to read.

File: StortSpelProjekt/Game/src/Gamefiles/Components/PlayerInputComponent.cpp [hunk 8]
```diff
@@ -194,14 +239,13 @@ void component::PlayerInputComponent::alternativeInput(ModifierInput* evnt)
...
-	m_CameraDistance = sqrt(m_Zoom * 4 * m_Zoom * 4 + m_Zoom * m_Zoom);
-}
-
 void component::PlayerInputComponent::move(MovementInput* evnt)
 {
+	if (evnt->pressed)
+	{
+		m_Attacking = false;
+	}
+
```
Comment (fjz345) [inline StortSpelProjekt/Game/src/Gamefiles/Components/PlayerInputComponent.cpp:268]: Attacking in a MovementInput Event function????

**Review by fjz345 — DISMISSED** [review]
reply to my changes about them.
I do not know what they've said in the game design but it felt very special to have a turn rate in a fps, might not be a bad thing though. The bad thing was the turning speed. Felt like I had no control at all of hitting the enemies.

**Review by Schtekt — DISMISSED** [review]
Looks good, the camera problem with walls is much more apparent now.

A bug that may not be your fault that I noted though is that the player continues to shoot projectiles if the pressed mouse button is switched from right button to left button during attack. Maybe something you can take a look at?

File: StortSpelProjekt/Engine/src/Renderer/Camera/BaseCamera.cpp (hunk 1 above)
Comment (LeoWikstrom) [reply StortSpelProjekt/Engine/src/Renderer/Camera/BaseCamera.cpp:11]: Nope. It pointed straight to the left before I changed this

File: StortSpelProjekt/Engine/src/Renderer/Camera/BaseCamera.cpp (hunk 2 above)
Comment (LeoWikstrom) [reply StortSpelProjekt/Engine/src/Renderer/Camera/BaseCamera.cpp:40]: Same as above

File: StortSpelProjekt/Engine/src/Renderer/Camera/PerspectiveCamera.cpp (hunk 3 above)
Comment (LeoWikstrom) [reply StortSpelProjekt/Engine/src/Renderer/Camera/PerspectiveCamera.cpp:89]: Same as above

File: StortSpelProjekt/Game/src/Gamefiles/Components/MeleeComponent.cpp (hunk 4 above)
Comment (LeoWikstrom) [reply StortSpelProjekt/Game/src/Gamefiles/Components/MeleeComponent.cpp:124]: You are wrong about that

File: StortSpelProjekt/Game/src/Gamefiles/Components/PlayerInputComponent.cpp (hunk 5 above)
Comment (LeoWikstrom) [reply StortSpelProjekt/Game/src/Gamefiles/Components/PlayerInputComponent.cpp:33]: Why? Makes no sense

File: StortSpelProjekt/Game/src/Gamefiles/Components/PlayerInputComponent.cpp (hunk 6 above)
Comment (LeoWikstrom) [reply StortSpelProjekt/Game/src/Gamefiles/Components/PlayerInputComponent.cpp:120]: That's fair. Will comment

File: StortSpelProjekt/Game/src/Gamefiles/Components/PlayerInputComponent.cpp (hunk 7 above)
Comment (LeoWikstrom) [reply StortSpelProjekt/Game/src/Gamefiles/Components/PlayerInputComponent.cpp]: Seems inefficient. Can comment and put in regions

File: StortSpelProjekt/Game/src/Gamefiles/Components/PlayerInputComponent.cpp (hunk 8 above)
Comment (LeoWikstrom) [reply StortSpelProjekt/Game/src/Gamefiles/Components/PlayerInputComponent.cpp:268]: No, set that player is not currently attacking so she will turn in the direction she is moving. Will add more comments 👍

File: StortSpelProjekt/Game/src/Gamefiles/Components/PlayerInputComponent.cpp (hunk 5 above)
Comment (fjz345) [reply StortSpelProjekt/Game/src/Gamefiles/Components/PlayerInputComponent.cpp:33]: Do not know what exactly PlayerInputComponents job is but attack cooldown is in meleecomponent, if an attack has not been done or if it has been done should be decided by meleecomponent in my understanding.

**Review by fjz345 — APPROVED** [review]

**Review by Schtekt — APPROVED** [review]

**Pull request merged**